src/psyclone/domain/common/transformations/kernel_module_inline_trans.py

```diff
@@ -83,34 +85,75 @@ def validate(self, node, options=None):
         :param options: a dictionary with options for transformations.
         :type options: Optional[Dict[str, Any]]
 
-        :raises TransformationError: if the target node is not a sub-class of \
-            psyGen.CodedKern.
-        :raises TransformationError: if the subroutine containing the \
-            implementation of the kernel cannot be retrieved with \
-            'get_kernel_schedule'.
-        :raises TransformationError: if the name of the routine that \
-            implements the kernel is not the same as the kernel name. This \
-            will happen if the kernel is polymorphic (uses a Fortran \
+        :raises TransformationError: if the target node is not a sub-class of
+            psyGen.CodedKern or psyir.nodes.Call or is an IntrinsicCall.
+        :raises TransformationError: if the target node is not within a
+            Container (Fortran module).
+        :raises TransformationError: if there is no explicit import of the
+            called Routine and there is already a Routine of that name in the
+            parent Container.
+        :raises TransformationError: if the PSyIR of the implementation of the
+            called Routine/kernel cannot be retrieved.
+        :raises TransformationError: if the name of the routine that
+            implements the kernel is not the same as the kernel name. This
+            will happen if the kernel is polymorphic (uses a Fortran
             INTERFACE) and will be resolved by #1824.
         :raises TransformationError: if the kernel cannot be safely inlined.
 
         '''
-        if not isinstance(node, CodedKern):
+        if isinstance(node, CodedKern):
+            routine_sym = None
+            kname = node.name
+            kern_or_call = "Kernel"
+        elif isinstance(node, Call):
+            if isinstance(node, IntrinsicCall):
+                raise TransformationError(
+                    f"Cannot module-inline a call to an intrinsic (got "
+                    f"'{node.debug_string()}')")
+            routine_sym = node.routine
+            kname = routine_sym.name
+            kern_or_call = "routine"
+        else:
```

Comment:
What's the plan of this (even if not implemented in this PR)?

If its supported for any generic calls it seems this transformation should be moved to src/psyclone/psyir/transformations but then this mentions CodedKern (which is not generic).  Some options are:
- do instead a specialisation of this transformation for the CodedKern (not ideal because we need to remember to import the specialised one when we are in the appropriate API)
- Make CodedKern inherit from Call and make kernel conform to the Call methods (there is an issue about this that I started a while ago but I never finished)

Also do you plan to drop the "Kernel" from the name of the transformation? And if we go through the pain of renaming, there has been some confusion in the past about the ambiguous meaning with "Inline a kernel module" instead of "Module Inline a Kernel", maybe we can improve the name to avoid this? (Should be Module or Container?, is Inline the right word for it?)

Reply:
My preference would be to make `CodedKern` inherit from `Call` and then we would drop the distinction here.
I hadn't thought about renaming but, yes, I agree it needs to happen. Perhaps it should be "ModuleInlineTrans" or, "ContainerInlineTrans"? I don't know what the "right" word is I'm afraid. We could go for "move" or "insert" perhaps so "InsertRoutineTrans" might be clearer as we're keeping the Routine, just moving it (as opposed to 'inlining' where the Routine disappears).

Reply:
What do @rupertford, @TeranIvy and @hiker think about the name we should use?

Reply:
@arporter Since we are bringing it into the local scope and the implementation is different/copied (changes here will not affect the other callers in other moduels) maybe it can use "localization" or "privatization" e.g "PrivatizeRoutineToLocalContainer"

Which made me think that whatever is called we need to mark the new symbol as "private", which we currently don't do and could potentially cause clashes from other modules importing the one where the caller is!